toponetx/utils/normalization.py

```diff
@@ -0,0 +1,383 @@
+    r_inv_sqrt[np.isinf(r_inv_sqrt)] = 0.0
+    r_mat_inv_sqrt = diags(r_inv_sqrt)
+    A_opt_normalized = A_opt.dot(r_mat_inv_sqrt).transpose().dot(r_mat_inv_sqrt)
+
+    return A_opt_normalized
+
+
+def compute_xu_asymmetric_normalized_matrix(B, is_sparse=True):
+    """
+    Compute Xu's normalized asymmetric matrix.
+
+    Typically used to normalize boundary operators.
+
+    Parameters
+    ----------
+    B : numpy.ndarray or scipy.sparse.csr_matrix
+        The asymmetric matrix.
+    is_sparse : bool, optional
+        If True, treat B as a sparse matrix.
+
+    Returns
+    -------
+    numpy.ndarray or scipy.sparse.csr_matrix
+        The normalized asymmetric matrix.
+
+    Notes
+    -----
+    This normalization is based on Xu's formulation, which computes the diagonal matrix D^(-1)
+    and multiplies it with the asymmetric matrix B.
+    """
+    if is_sparse:
+        rowsum = np.array(np.abs(B).sum(1))
+        r_inv = np.power(rowsum, -1).flatten()
+        r_inv[np.isinf(r_inv)] = 0.0
+        r_mat_inv = diags(r_inv)
+        normalized_matrix = r_mat_inv.dot(B)
+        return normalized_matrix
+    else:
+        Di = np.sum(np.abs(B), axis=1)
+        Di2 = np.diag(np.power(Di, -1))
+        Di2[np.isinf(Di2)] = 0.0
+        B = np.array(B)
+        normalized_matrix = Di2.dot(B)
+        return normalized_matrix
+
+
+def compute_bunch_normalized_matrices(B1, B2):
+    """Get Bunch normalization.
+
+    Parameters
+    ----------
+    B1 : numpy array or scipy csr_matrix
+        The boundary B1: C1->C0 of a complex.
+    B2 : numpy array or scipy csr_matrix
+        The boundary B2: C2->C1 of a complex.
+
+    Returns
+    -------
+    B1 : numpy array or scipy csr_matrix
+        Normalized B1: C1->C0.
+    B1T : numpy array or scipy csr_matrix
+        Normalized B1T: C0->C1.
+    B2 : numpy array or scipy csr_matrix
+        Normalized B2: C2->C1.
+    B2T : numpy array or scipy csr_matrix
+        Normalized B2T: C1->C2.
+
+    References
+    ----------
+    [1] Michael T. Schaub, Austin R. Benson,
+        Paul Horn, Gabor Lippner,
+        Ali Jadbabaie Random walks on
+        simplicial complexes and the normalized
+        hodge 1-laplacian.
+    [2] Eric Bunch, Qian You,
+        Glenn Fung, Vikas Singh
+        Simplicial 2-Complex Convolutional
+        Neural Networks
+    """
+    B1N = _compute_B1_normalized_matrix(B1, B2)
+    B1TN = _compute_B1T_normalized_matrix(B1, B2)
+    B2N = _compute_B2_normalized_matrix(B2)
+    B2TN = _compute_B2T_normalized_matrix(B2)
+    return B1N, B1TN, B2N, B2TN
+
+
+def _compute_B1_normalized_matrix(B1, B2):
+    """Compute normalized boundary matrix B1.
+
+    Parameters
+    ----------
+    B1 : numpy array or scipy csr_matrix
+        The boundary B1: C1->C0 of a complex.
+    B2 : numpy array or scipy csr_matrix
+        The boundary B2: C2->C1 of a complex.
+
+    Returns
+    -------
+    _ : numpy array or scipy csr_matrix
```

Comment:
Please remove the `_ : ` here and below. If a function returns just one value (as opposed to a tuple of values), just state the type.